hugr-model/src/v0/mod.rs

```diff
@@ -0,0 +1,368 @@
+//! are treated uniformly as terms, enabling a restricted form of dependent typing.
+//! The type system is extensible and can be used to declaratively encode the desired shape
+//! of operation parameters and metadata. Type constraints can be used to express more complex
+//! validation rules.
+//!
+//! # Tabling
+//!
+//! Instead of directly nesting structures, we store them in tables and refer to them
+//! by their index in the table. This allows us to attach additional data to the structures
+//! without changing the data structure itself. This can be used, for example, to keep track
+//! of metadata that has been parsed from its generic representation as a term into a more
+//! specific in-memory representation.
+//!
+//! The tabling is also used for deduplication of terms. In practice, many terms will share
+//! the same subterms, and we can save memory and validation time by storing them only once.
+//! However we allow non-deduplicated terms for cases in which terms carry additional identity
+//! over just their structure. For instance, structurally identical terms could originate
+//! from different locations in a text file and therefore should be treated differently when
+//! locating type errors.
+use smol_str::SmolStr;
+use tinyvec::TinyVec;
+
+/// Index of a node in a hugr graph.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct NodeId(pub u32);
+
+/// Index of a port in a hugr graph.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct PortId(pub u32);
+
+/// Index of an edge in a hugr graph.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct EdgeId(pub u32);
+
+/// Index of a term in a hugr graph.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+pub struct TermId(pub u32);
+
+/// An identifier referring to types, terms, or functions.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Symbol(pub SmolStr);
+
+/// A local variable in terms.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TermVar(pub SmolStr);
+
+/// The name of an edge.
+///
+/// This is to be used in the textual representation of the graph
+/// to indicate that two ports are connected by assigning them the same
+/// edge variable.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct EdgeVar(pub SmolStr);
+
+/// A module consisting of a hugr graph together with terms.
+#[derive(Debug, Clone, Default, PartialEq, Eq, Hash)]
+pub struct Module {
+    /// Table of [`Node`]s.
+    pub nodes: Vec<Node>,
+    /// Table of [`Port`]s.
+    pub ports: Vec<Port>,
+    /// Table of [`Term`]s.
+    pub terms: Vec<Term>,
+}
+
+/// Nodes in the hugr graph.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Node {
+    /// The operation that the node performs.
+    pub operation: Operation,
+    /// Parameters that are passed to the operation.
```

Comment:
This is an arbitrary-length (presumed small) list of triples of TermId? Please explain here why it's a triple i.e. what each of the three elements are.

Reply:
It's a vector that stores its elements inline (without a heap allocation) if it is small enough. This has no semantic significance but is a simple optimisation. `TinyVec<[u32; 3]>` is the same size as `Vec<u32>`. `TinyVec<[u32; 4]>` is bigger than `Vec<u32>`.

Reply:
oh, the `;3` is like the maximum size to store inline? Ok, right